sound/soc/sof/sof-audio.c

```diff
@@ -85,16 +86,20 @@ static int sof_widget_free_unlocked(struct snd_sof_dev *sdev,
 			err = ret;
 	}
 
-	/*
-	 * disable widget core. continue to route setup status and complete flag
-	 * even if this fails and return the appropriate error
-	 */
-	ret = snd_sof_dsp_core_put(sdev, swidget->core);
-	if (ret < 0) {
-		dev_err(sdev->dev, "error: failed to disable target core: %d for widget %s\n",
-			swidget->core, swidget->widget->name);
-		if (!err)
-			err = ret;
+	/* disable cores associated with all modules in the pipeline and clear the complete flag */
+	if (swidget->id == snd_soc_dapm_scheduler) {
+		int i;
+
+		for_each_set_bit(i, &spipe->core_mask, sdev->num_cores) {
+			ret = snd_sof_dsp_core_put(sdev, i);
```

Comment:
the only part that scares is the locking model. How is the refcount protected between get and put?

Reply:
@plbossart widget setup/free is protected by the widget setup_mutex and also the dpcm mutex during hw_params/free by the core. We've never had a mutex protecting core ref counts before. Im not sure it is really needed.

Reply:
well the problem is that the dpcm mutex only applies between hw_params and hw_free *of the same PCM device*. But here there is no real correlation when you have shared BE pipelines inside the firmware. With the deferred count I think you have a larger attack surface to worry about, no?

Reply:
with shared BE pipelines, we have the widget_mutex protecting the widget setup/free no?